File: frontend/src/types/websocket.ts
/**
 * WebSocket message protocol types matching backend
 */

/**
 * System statistics update message
 */
export interface WSStatsUpdate {
  type: "stats_update";
  cpu_percent: number;
  memory_percent: number;
  disk_free_gb: number;
  temperature_celsius: number | null;
  timestamp: string;
}

/**
 * Camera event message
 */
export interface WSCameraEvent {
  type: "camera_event";
  camera_id: number;
  event: "online" | "offline" | "recording_started" | "recording_stopped" | "error";
  message?: string | null;
  timestamp: string;
}

/**
 * Job update message
 */
export interface WSJobUpdate {
  type: "job_update";
  job_id: number;
  camera_id: number;
  job_type: string;
  status: "pending" | "running" | "completed" | "failed" | "interrupted";
  progress?: number | null; // 0-100
  timestamp: string;
}

/**
 * Error message
 */
export interface WSError {
  type: "error";
  code: string;
  message: string;
  timestamp: string;
}

/**
 * Union type of all possible WebSocket messages
 */
export type WSMessage = WSStatsUpdate | WSCameraEvent | WSJobUpdate | WSError;

/**
 * Type guards for message discrimination
 */
export function isStatsUpdate(msg: WSMessage): msg is WSStatsUpdate {
  return msg.type === "stats_update";
}

export function isCameraEvent(msg: WSMessage): msg is WSCameraEvent {
  return msg.type === "camera_event";
}

export function isJobUpdate(msg: WSMessage): msg is WSJobUpdate {
  return msg.type === "job_update";
}

export function isError(msg: WSMessage): msg is WSError {
  return msg.type === "error";
}

/**
 * WebSocket message handler type
 */
export type MessageHandler<T extends WSMessage = WSMessage> = (message: T) => void;

/**
 * WebSocket connection state
 */
export type ConnectionState =
  | "disconnected"
  | "connecting"
  | "connected"
  | "reconnecting"
  | "error";

File: backend/app/models/schemas/websocket.py
"""WebSocket message protocol schemas."""

from datetime import datetime
from typing import Literal, Optional, Union

from pydantic import BaseModel, Field


class WSStatsUpdate(BaseModel):
    """System statistics update message."""

    type: Literal["stats_update"] = "stats_update"
    cpu_percent: float = Field(..., description="CPU usage percentage")
    memory_percent: float = Field(..., description="Memory usage percentage")
    disk_free_gb: float = Field(..., description="Free disk space in GB")
    temperature_celsius: Optional[float] = Field(
        None, description="CPU temperature in Celsius"
    )
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class WSCameraEvent(BaseModel):
    """Camera status event message."""

    type: Literal["camera_event"] = "camera_event"
    camera_id: int = Field(..., description="Camera database ID")
    event: Literal[
        "online", "offline", "recording_started", "recording_stopped", "error"
    ] = Field(..., description="Event type")
    message: Optional[str] = Field(None, description="Additional event message")
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class WSJobUpdate(BaseModel):
    """Job progress update message."""

    type: Literal["job_update"] = "job_update"
    job_id: int = Field(..., description="Job database ID")
    camera_id: int = Field(..., description="Camera database ID")
    job_type: str = Field(..., description="Job type (capture, record, timelapse)")
    status: Literal["pending", "running", "completed", "failed", "interrupted"] = Field(
        ..., description="Job status"
    )
    progress: Optional[float] = Field(
        None, ge=0, le=100, description="Progress percentage (0-100)"
    )
    current_frame: Optional[int] = Field(
        None, ge=0, description="Current frame count (for timelapse jobs)"
    )
    total_frames: Optional[int] = Field(
        None, ge=0, description="Total expected frames (for timelapse jobs, None=unlimited)"
    )
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class WSError(BaseModel):
    """Error message."""

    type: Literal["error"] = "error"
    code: str = Field(..., description="Error code")
    message: str = Field(..., description="Error message")
    timestamp: datetime = Field(default_factory=datetime.utcnow)


# Union type for all WebSocket messages
WSMessage = Union[WSStatsUpdate, WSCameraEvent, WSJobUpdate, WSError]
